code/dA.py

```diff
@@ -336,7 +336,7 @@ def test_dA(learning_rate=0.1, training_epochs=15,
         for batch_index in range(n_train_batches):
             c.append(train_da(batch_index))
 
-        print('Training epoch %d, cost ' % epoch, numpy.mean(c))
+        print('Training epoch %d, cost ' % epoch, numpy.mean(c, dtype='float64'))
```

Comment:
Do you know what is going on about numpy.mean? if c are float16 output I checked that numpy output float16. In my tests, it seem the accumulator is in float16 or something like that. Do you know?

We would need to document that in Theano about float16.  At least in this issue: https://github.com/Theano/Theano/issues/2908. I let you modify it, in case you can add more information.

Should we special case float16 and make Theano always return at least float32 to help prevent that type of problems?